Update for breaking API changes in MuPDF 1.14

Compilation errors under MuPDF 1.14:

```
g++ -Wall -O2 -std=c++1y -DMUPDF_VERSION=10014   -c -o pdf_document.o pdf_document.cpp                                                pdf_document.cpp: In member function ‘virtual void PDFDocument::Render(Document::PixelWriter*, int, float, int)’:
pdf_document.cpp:169:48: error: could not convert ‘& bbox’ from ‘const fz_irect*’ {aka ‘const fz_irect_s*’} to ‘fz_irect’ {aka ‘fz_irect_s’}
       _fz_context, fz_device_rgb(_fz_context), &bbox, nullptr, 1);                                                                                                                   ^~~~~
pdf_document.cpp:170:52: error: could not convert ‘& fz_identity’ from ‘const fz_matrix*’ {aka ‘const fz_matrix_s*’} to ‘fz_matrix’ {aka ‘fz_matrix_s’}
   fz_device* dev = fz_new_draw_device(_fz_context, &fz_identity, pixmap);                                                                                                                ^~~~~~~~~~~~
pdf_document.cpp:174:62: error: could not convert ‘& m’ from ‘const fz_matrix*’ {aka ‘const fz_matrix_s*’} to ‘fz_matrix’ {aka ‘fz_matrix_s’}
pdf_document.cpp:51:47:
        pdf_run_page((context), (page), (dev), (matrix), (cookie))
                                               ~~~~~~~~
pdf_document.cpp:174:62:
   pdf_run_page(_fz_context, _pdf_document, page_struct, dev, &m, nullptr);
pdf_document.cpp:51:48: note: in definition of macro ‘pdf_run_page’
        pdf_run_page((context), (page), (dev), (matrix), (cookie))
                                                ^~~~~~
pdf_document.cpp: In member function ‘fz_matrix PDFDocument::Transform(float, int)’:
pdf_document.cpp:410:12: error: cannot convert ‘fz_matrix*’ {aka ‘fz_matrix_s*’} to ‘float’
   fz_scale(&scale_matrix, zoom, zoom);
            ^~~~~~~~~~~~~
In file included from /usr/include/mupdf/fitz/context.h:6,
                 from /usr/include/mupdf/fitz.h:11,
                 from /usr/include/mupdf/pdf.h:4,
                 from pdf_document.cpp:29:
/usr/include/mupdf/fitz/geometry.h:305:26: note:   initializing argument 1 of ‘fz_matrix fz_scale(float, float)’
 fz_matrix fz_scale(float sx, float sy);
                    ~~~~~~^~
pdf_document.cpp:411:13: error: cannot convert ‘fz_matrix*’ {aka ‘fz_matrix_s*’} to ‘float’
   fz_rotate(&rotate_matrix, rotation);
             ^~~~~~~~~~~~~~
In file included from /usr/include/mupdf/fitz/context.h:6,
                 from /usr/include/mupdf/fitz.h:11,
                 from /usr/include/mupdf/pdf.h:4,
                 from pdf_document.cpp:29:
/usr/include/mupdf/fitz/geometry.h:374:27: note:   initializing argument 1 of ‘fz_matrix fz_rotate(float)’
 fz_matrix fz_rotate(float degrees);                                                                                                                       ~~~~~~^~~~~~~
pdf_document.cpp:412:13: error: could not convert ‘& transformation_matrix’ from ‘fz_matrix*’ {aka ‘fz_matrix_s*’} to ‘fz_matrix’ {aka ‘fz_matrix_s’}
   fz_concat(&transformation_matrix, &scale_matrix, & rotate_matrix);
             ^~~~~~~~~~~~~~~~~~~~~~
pdf_document.cpp: In member function ‘fz_irect PDFDocument::GetBoundingBox(pdf_page*, const fz_matrix&)’:
pdf_document.cpp:53:48: error: too many arguments to function ‘fz_rect pdf_bound_page(fz_context*, pdf_page*)’
        pdf_bound_page((context), (page), (bbox))
                                                ^
pdf_document.cpp:422:7: note: in expansion of macro ‘pdf_bound_page’
       pdf_bound_page(_fz_context, _pdf_document, page_struct, &bbox), &m));
       ^~~~~~~~~~~~~~
In file included from /usr/include/mupdf/pdf.h:16,
                 from pdf_document.cpp:29:
/usr/include/mupdf/pdf/page.h:61:9: note: declared here
 fz_rect pdf_bound_page(fz_context *ctx, pdf_page *page);                                                                                      ^~~~~~~~~~~~~~
pdf_document.cpp:422:71: error: could not convert ‘& m’ from ‘const fz_matrix*’ {aka ‘const fz_matrix_s*’} to ‘fz_matrix’ {aka ‘fz_matrix_s’}
       pdf_bound_page(_fz_context, _pdf_document, page_struct, &bbox), &m));
                                                                       ^~
make: *** [<builtin>: pdf_document.o] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -47,6 +47,11 @@ LIBS := \
     -ljbig2dec \
     -ljpeg \
     -l$(OPENJP2)
+ifeq ($(shell [ $(MUPDF_VERSION) -ge 10014 ]; echo $$?), 0)
+  LIBS += -lmupdf-third \
+      -lssl \
+      -lcrypto
+else
 ifeq ($(shell [ $(MUPDF_VERSION) -ge 10009 ]; echo $$?), 0)
   LIBS += -lmupdfthird \
       -lssl \
@@ -58,6 +63,7 @@ ifeq ($(shell [ $(MUPDF_VERSION) -ge 10004 ]; echo $$?), 0)
       -lcrypto
 else
   LIBS += -lmupdf-js-none
+endif
 endif
 endif
 

--- a/pdf_document.cpp
+++ b/pdf_document.cpp
@@ -46,30 +46,38 @@ extern "C" {
 #define fz_drop_stext_page fz_drop_text_page
 #endif
 
+#if MUPDF_VERSION < 10007
+#define pdf_bound_page(context, pdf_document, page, bbox) \
+  pdf_bound_page((pdf_document), (page), (bbox))
+#elif MUPDF_VERSION < 10014
+#define pdf_bound_page(context, pdf_document, page, bbox) \
+  pdf_bound_page((context), (page), (bbox))
+#else
+// No definition
+#endif
+
+#if MUPDF_VERSION < 10007
+#define pdf_drop_page(context, pdf_document, page) \
+  pdf_free_page((pdf_document), (page))
+#elif MUPDF_VERSION < 10011
+#define pdf_drop_page(context, pdf_document, page) \
+  pdf_drop_page((context), (page))
+#else
+#define pdf_drop_page(context, pdf_document, page) \
+  fz_drop_page((context), &(page)->super)
+#endif
+
 #if MUPDF_VERSION >= 10007
 #define pdf_run_page(context, pdf_document, page, dev, matrix, cookie) \
   pdf_run_page((context), (page), (dev), (matrix), (cookie))
-#define pdf_bound_page(context, pdf_document, page, bbox) \
-  pdf_bound_page((context), (page), (bbox))
-#if MUPDF_VERSION >= 10011
-#define pdf_drop_page(context, pdf_document, page) \
-  fz_drop_page((context), &(page)->super)
-#else
-#define pdf_drop_page(context, pdf_document, page) \
-  pdf_drop_page((context), (page))
-#endif
 #else
 #define pdf_run_page(context, pdf_document, page, dev, matrix, cookie) \
   pdf_run_page((pdf_document), (page), (dev), (matrix), (cookie))
-#define pdf_bound_page(context, pdf_document, page, bbox) \
-  pdf_bound_page((pdf_document), (page), (bbox))
 #define fz_drop_context fz_free_context
 #define fz_drop_device(context, dev) fz_free_device(dev)
 #define fz_drop_text_page fz_free_text_page
 #define fz_drop_text_sheet fz_free_text_sheet
 #define fz_drop_outline fz_free_outline
-#define pdf_drop_page(context, pdf_document, page) \
-  pdf_free_page((pdf_document), (page))
 #define fz_begin_page(context, dev, rect, identity) \
   (fz_begin_page((dev), (rect), (identity)))
 #define fz_end_page(context, dev) (fz_end_page(dev))
@@ -93,6 +101,12 @@ extern "C" {
 #define fz_new_draw_device(context, transform, dest) \
   (fz_new_draw_device(context, dest))
 #define fz_close_device(context, dev)
+#endif
+
+#if MUPDF_VERSION < 10014
+#define FZ_OBJ(x) (&(x))
+#else
+#define FZ_OBJ(x) (x)
 #endif
 
 const char* const PDFDocument::DEFAULT_ROOT_OUTLINE_ITEM_TITLE =
@@ -160,12 +174,13 @@ void PDFDocument::Render(
   pdf_page* page_struct = GetPage(page);
   const fz_irect& bbox = GetBoundingBox(page_struct, m);
   fz_pixmap* pixmap = fz_new_pixmap_with_bbox(
-      _fz_context, fz_device_rgb(_fz_context), &bbox, nullptr, 1);
-  fz_device* dev = fz_new_draw_device(_fz_context, &fz_identity, pixmap);
+      _fz_context, fz_device_rgb(_fz_context), FZ_OBJ(bbox), nullptr, 1);
+  fz_device* dev = fz_new_draw_device(_fz_context, FZ_OBJ(fz_identity), pixmap);
 
   // 2. Render page.
   fz_clear_pixmap_with_value(_fz_context, pixmap, 0xff);
-  pdf_run_page(_fz_context, _pdf_document, page_struct, dev, &m, nullptr);
+  pdf_run_page(
+      _fz_context, _pdf_document, page_struct, dev, FZ_OBJ(m), nullptr);
 
   // 3. Write pixmap to buffer. The page is vertically divided into n equal
   // stripes, each copied to pw by one thread.
@@ -387,19 +402,31 @@ pdf_page* PDFDocument::GetPage(int page) {
 
 fz_matrix PDFDocument::Transform(float zoom, int rotation) {
   fz_matrix transformation_matrix, scale_matrix, rotate_matrix;
+#if MUPDF_VERSION >= 10014
+  scale_matrix = fz_scale(zoom, zoom);
+  rotate_matrix = fz_rotate(rotation);
+  transformation_matrix = fz_concat(scale_matrix, rotate_matrix);
+#else
   fz_scale(&scale_matrix, zoom, zoom);
   fz_rotate(&rotate_matrix, rotation);
   fz_concat(&transformation_matrix, &scale_matrix, &rotate_matrix);
+#endif
+
   return transformation_matrix;
 }
 
 fz_irect PDFDocument::GetBoundingBox(
     pdf_page* page_struct, const fz_matrix& m) {
   assert(page_struct != nullptr);
+#if MUPDF_VERSION >= 10014
+  return fz_round_rect(
+      fz_transform_rect(pdf_bound_page(_fz_context, page_struct), m));
+#else
   fz_rect bbox;
   fz_irect ibbox;
   return *fz_round_rect(
       &ibbox,
       fz_transform_rect(
           pdf_bound_page(_fz_context, _pdf_document, page_struct, &bbox), &m));
+#endif
 }

